src/Raven.Server/Documents/Expiration/ExpirationStorage.cs

```diff
@@ -146,38 +159,90 @@ private void ThrowWrongExpirationDateFormat(Slice lowerId, string expirationDate
+                {
+                    if (currentTime >= date.ToUniversalTime())
+                        return true;
+                }
+            }
+            return false;
         }
 
         public int DeleteExpiredDocuments(DocumentsOperationContext context, Dictionary<Slice, List<(Slice LowerId, LazyStringValue Id)>> expired)
         {
             var deletionCount = 0;
             var expirationTree = context.Transaction.InnerTransaction.ReadTree(DocumentsByExpiration);
+            var currentTime = _database.Time.GetUtcNow();
 
             foreach (var pair in expired)
             {
                 foreach (var ids in pair.Value)
                 {
                     if (ids.Id != null)
                     {
-                        var deleted = _database.DocumentsStorage.Delete(context, ids.LowerId, ids.Id, expectedChangeVector: null);
-
-                        if (_logger.IsInfoEnabled && deleted == null)
-                            _logger.Info($"Tried to delete expired document '{ids.Id}' but document was not found.");
+                        try
+                        {
+                            var doc = _database.DocumentsStorage.Get(context, ids.LowerId, throwOnConflict: true);
+                            if (doc != null && doc.TryGetMetadata(out var metadata))
+                            {
+                                if (HasExpired(metadata, Constants.Documents.Metadata.Expires, currentTime))
+                                {
+                                    _database.DocumentsStorage.Delete(context, ids.LowerId, ids.Id, expectedChangeVector: null);
+                                }
+                                else if (HasExpired(metadata, Constants.Documents.Metadata.Refresh, currentTime))
+                                {
+                                    // remove the @refresh tag
+                                    metadata.Modifications = new Sparrow.Json.Parsing.DynamicJsonValue(metadata);
+                                    metadata.Modifications.Remove(Constants.Documents.Metadata.Refresh);
+                                    
+                                    using(var updated = context.ReadObject(doc.Data, doc.Id, BlittableJsonDocumentBuilder.UsageMode.ToDisk))
+                                    {
+                                        _database.DocumentsStorage.Put(context, doc.Id, doc.ChangeVector, updated, flags: doc.Flags);
+                                    }
+                                }
+                            }
+                        }
+                        catch (DocumentConflictException)
+                        {
+                            if(GetConflictedExpiration(context, currentTime, ids.LowerId).AllExpired)
+                                _database.DocumentsStorage.Delete(context, ids.LowerId, ids.Id, expectedChangeVector: null);
```

Comment:
i'm not sure if this behavior is right if we are running on '@refresh' and get a conflict shouldn't we do a Put here rather than a delete?

Reply:
This can only happen if all the conflicts have expired, so we can auto resolve by delete.
Can't happen if there isn't such a scenario, so I don't think there is a problem here?